.changeset/long-insects-remember.md

```diff
@@ -0,0 +1,7 @@
+---
+'xstate': patch
+---
+
+Fixed type compatibility with functions accepting machines that were created before typegen was a thing in XState. This should make it possible to use the latest version of XState with `@xstate/vue`, `@xstate/react@^1` and some community packages.
+
+Note that this change doesn't make those functions to accept machines that have typegen information on them. For that the signatures of those functions would have to be adjusted.
```

Comment:
This isn't possible because we are using `__TResolvedTypesMeta` phantom type [here](https://github.com/statelyai/xstate/blob/e75c47ad147453aa5ba7dd7723c21d30ef6af44f/packages/core/src/types.ts#L1067) and this makes a machine with typegen information incompatible with a machine that has typegen information disabled for it.

**Maybe, maybe** it would be possible to fix this by defaulting to `any` and pushing resolving logic for the `TypegenDisabled` case further down the road. This would require using `IsAny` and stuff like that and, in general, I've experienced `any`s wreaking havoc on the type machine. Maybe it can be wrangled but I'm somewhat scared of doing such a substantial change now cause while this type machinery works quite alright... it's somewhat fragile.

I think that, for the time being, this is a reasonable tradeoff for what the typegen enables us to do. It feels better to push it harder than to experiment with changes that might break it. I imagine that with some of the changes planned for v5 this might become much less of a problem - but also I can't be certain about it until we actually try to minimize the amount of generics etc.

Reply:
Do we want to create an `AnyTypegenMachine` type to help with this?

Reply:
As mentioned in the private conversation - this wouldn't work correctly because the whole type machinery is built on top of a type that has some specific structure. `ResolvedTypegenDisabled` is acting, sort of, like `UnknownTypegenMeta`. What I mean by this is that the typegen meta includes string literals etc when the `ResolvedTypegenDisabled` has just `string`, `Record<string, unknown>` etc. In a way probably this holds true:
```ts
const truthy: ResolvedTypegenMeta extends ResolvedTypegenDisabled ? true : false = true
```
This allows the whole machine after the initial resolution to work exactly the same - for the most part, I don't even need to check if the resolved type contains metadata created by the typegen or not, I don't branch the implementation based on this.

This is just part of the problem though. The first problem that would have to be addressed before even thinking about any kind of reactors here is related to `machine.withConfig`. This method utilizes the typegen information and might mark some implementations as **required**. If the consuming function doesn't specify the typegen slot anyhow we need to default it to something. Whatever we default it to though has to end up marking all implementations provided to `withConfig` as optional - because this is like the most "loose" shape.

So the consuming function that doesn't specify the typegen slot anyhow will always end up expecting a machine that has `.withConfig(options: EverythingOptional)` while the typegen-based provides something like `.withConfig(options: SomeMightBeRequired)`. Those types are just fundamentally incompatible because allowing this would be unsound, especially when we consider `strictFunctionTypes` (and I believe that it's our target to support this option, in a way - I don't intend to work on any special support for the opposite).

So it feels like the only viable solution here is to enforce the choice on those consuming functions:
1. they can either provide `any` for the typegen slot to say "I don't care about this as I don't do anything special with it"
2. or they can infer it or provide a generic for it to utilize the information or forward it to the result of their function or something